Updating reauthorizationRequired notifications (#15533)
microsoftgraph/microsoft-graph-docs: concepts/webhooks-lifecycle.md

When you receive a `reauthorizationRequired` lifecycle notification, you must reauthorize the subscription to maintain the data flow.

> You can create a long-lived subscription depending on the [maximum supported subscription life time for the resource](/graph/api/resources/subscription#maximum-length-of-subscription-per-resource-type), which enables change notifications to flow to the **notificationUrl**. If the conditions of access have changed since the subscription was created or Microsoft Graph detects that the notification flow might be inturrupted in the near future, Microsoft Graph may require that you reauthorize the subscription to prove that you still have access to resource data. The following are examples of conditions that may affect your access to data:

Reviewer: ```suggestion
You can create a long-lived subscription depending on the [maximum supported subscription lifetime for the resource](/graph/api/resources/subscription#maximum-length-of-subscription-per-resource-type), which enables change notifications to flow to the **notificationUrl**. If the conditions of access changed after the subscription was created, or Microsoft Graph detects that the notification flow might be interrupted in the near future, Microsoft Graph may require that you reauthorize the subscription to prove that you still have access to the resource's data. The following are examples of conditions that may affect your access to data:
```

- A tenant administrator may revoke your app's permissions to read a resource.
- In an interactive scenario, the user who provides the authentication token to your app may be subject to dynamic policies based on various factors, such as their location, device state, or risk assessment. For example, if the user changes their physical location, the user may no longer be allowed to access the data, and your app will not be able to reauthorize the subscription. For more information about dynamic policies that control access, see [Azure AD conditional access policies](/azure/active-directory/conditional-access/overview). 

> - Your access token expires. This only applies to notifications which include resource data
> - The subscription expires before you renew it

Reviewer: ```suggestion
- Your access token expires. This only applies to notifications that include resource data.
- The subscription expires before you renew it.
```

> Before any of these conditions become true, Microsoft Graph will send an authorization challenge to the **lifecycleNotificationUrl**. The interval of these notifications is illustrated below:

Reviewer: I'm assuming that this is intended for the client?
```suggestion
Before any of these conditions become true, Microsoft Graph will send an authorization challenge to the **lifecycleNotificationUrl**. The following code sample illustrates how to configure intervals when Microsoft Graph sends the reauthorization notifications:
```

```csharp

>     var TokenTimeToExpirationInMinutes=(TokenExpirationTime-CurrentTime)/4;

Reviewer: ```suggestion
// The following code is for illustrative purposes only.

    var TokenTimeToExpirationInMinutes=(TokenExpirationTime-CurrentTime)/4;
```

    Note that the flow of change notifications may continue for a while, giving you extra time to respond. However, eventually change notification delivery pauses, until you take the required action.

3. Respond to this lifecycle notification in one of two ways:
    - Reauthorize the subscription. This does not extend the expiration date of the subscription.

>     - Renew the subscription. This both reauthorizes and extends the expiration date.

Reviewer: ```suggestion
    - Renew the subscription. This reauthorizes and extends the expiration date.
```

    Note: Both actions require you to present a valid authentication token, similar to [creating a new subscription](webhooks.md#creating-a-subscription) or [renewing a subscription before its expiration](webhooks.md#renewing-a-subscription).

> 4. If you successfully reauthorize or renew the subscription, change notifications continue. Otherwise, change notifications remain paused. Please note that Microsoft graph will drop the notifications at 4 hour mark after they are paused.

Reviewer: ```suggestion
4. If you successfully reauthorize or renew the subscription, change notifications continue. Otherwise, change notifications remain paused. Please note that Microsoft graph will drop the notifications four hours after they are paused.
```

A few things to note about this type of notification:

> - The `"lifecycleEvent": "reauthorizationRequired"` field designates this notification as an authorization challenge. Other types of lifecycle notifications are also possible <provide link for missed notifications and subscriptionRemoved notifications>, and new ones will be introduced in the future.

Reviewer: We'll deal with the future when it comes.
```suggestion
- The `"lifecycleEvent": "reauthorizationRequired"` field designates this notification as an authorization challenge. `missed` and `subscriptionRemoved` **lifecycleEvent** notifications are also supported.
```
